[Round3] Step 1: 타입과 인스턴스의 관계

ollekil opened the pull request:
@jaemuYeo 잼킹님 pr 남깁니다.

. main파일
1. Person 인스턴스 생성
2. CoffeShop 인스턴스 생성
3. yagombucks의 바리스타를 misterLee로 할당
4. missKim이 커피를 구매함
5. misterLee가 커피를 구매함
6. missKim이 커피를 한잔 더 구매함
7. 결과값 프린트 출력
8. 커피숍 매출액 프린트 출력

. Coffe 파일
1. 커피종류 열거형 정의

. Person 파일
1. 사람에 관한 타입 정의 클래스
2. 커피 구매 함수

. CoffeShop 파일
1. 커피숍 이름, 매출액, 메뉴판, 주문한 커피테이블 저장, 바리스타 프로퍼티 생성
2. 커피를 만들어서 저장하는 함수
3. 바리스타 저장하는 함수

ollekil committed:
. main파일
1. Person 인스턴스 생성
2. CoffeShop 인스턴스 생성
3. yagombucks의 바리스타를 misterLee로 할당
4. missKim이 커피를 구매함
5. misterLee가 커피를 구매함
6. missKim이 커피를 한잔 더 구매함
7. 결과값 프린트 출력
8. 커피숍 매출액 프린트 출력

. Coffe 파일
1. 커피종류 열거형 정의

. Person 파일
1. 사람에 관한 타입 정의 클래스
2. 커피 구매 함수

. CoffeShop 파일
1. 커피숍 이름, 매출액, 메뉴판, 주문한 커피테이블 저장, 바리스타 프로퍼티 생성
2. 커피를 만들어서 저장하는 함수
3. 바리스타 저장하는 함수

jaemuYeo commented on CodeStarterCamp_Week3/Step1/Class/CoffeeShop.swift:
```diff
@@ -0,0 +1,34 @@
+        self.pickUpTable = []
+    }
+    
+    func makeCoffe(coffee: Coffee) {
```
파라미터 네임과 함수의 네임에 겹치는 것이 있네요!
어떻게 수정해보면 좋을까요?

ollekil replied:
커피의 종류이니까 coffeeType 으로 수정 하였습니다.

ollekil replied:
아래 buy 함수 처럼 
makeCoffee -> make 로 다시 바꿨습니다.

jaemuYeo commented on CodeStarterCamp_Week3/Step1/Class/CoffeeShop.swift:
```diff
@@ -0,0 +1,34 @@
+class CoffeeShop {
+    var name: String                // 카페 이름
+    var revenue: Double             // 매출액
+    var menu: [Coffee: Double]      // 메뉴판(커피종류key, 가격value)
```
커피의 가격을 Double타입으로 지정한 이유는 무엇인가요?

ollekil replied:
아.. 소수점으로 값을 표시 한번 해보고 싶어 double 을 썼었습니다. 
일반적인 원 개념으로 바꿔서 Double -> Int 수정 했습니다.

jaemuYeo commented on CodeStarterCamp_Week3/Step1/Class/CoffeeShop.swift:
```diff
@@ -0,0 +1,34 @@
+import Foundation
+
+// CoffeeShop 타입 정의
+class CoffeeShop {
```
CoffeeShop 타입이 클래스로 정의된 이유가 궁금합니다!

ollekil replied:
CoffeeShop 을 구조체로 정의를 하면 구조체는 값을 복사하는 타입이라 
커피숍에 수익이라던지 픽업테이블 데이터를 공유할수 없습니다. 
클래스는 값의 주소를 바라보는 참조타입이라 데이터를 공유 할수 있습니다. 
그래서 구조체보단 클래스가 적합하다 생각 했습니다. 
추후 상속등 확장성에서도 class 가 유용합니다.

jaemuYeo commented on CodeStarterCamp_Week3/Step1/Class/CoffeeShop.swift:
```diff
@@ -0,0 +1,34 @@
+        print("\(coffee.rawValue)가 나왔습니다. pickup 하세요.")
+    }
+    
+    func setBarista(barista: Person) {
```
오우! 바리스타가 누구인지 프린트하도록 함수도 만들어보셨군요 👍
기존에 의도한 부분은 yagombucks의 인스턴스를 생성할때 바로 주입해주는 것이였는데
새로운 방법인 것 같아요 ~

ollekil replied:
커피숍 클래스에서 가지고 있는 정보는 어떤 커피숍의 어떤 바리스타가 어떤 일을 하는지 알수 있기 때문에 
그 정보들을 출력 했습니다.

jaemuYeo commented on CodeStarterCamp_Week3/Step1/Class/Person.swift:
```diff
@@ -0,0 +1,31 @@
+import Foundation
+
+// Person 타입 정의
+class Person {
```
Person타입이 클래스로 정의된 이유도 궁금합니다!

ollekil replied:
Person 은 현재로선 Struct 구조체로 써도 무방해 보입니다. 
다만 추후 Person 은 Person 을 상속받은 ollekil로도 확장 할수 있는 여지가 있는 인스턴스이고.
Person 이 부모라 가정하고 ollekil, bjk 가 자식이고 부모의 돈을 공유 한다고 생각 했을때도 구조체로 하기보다 클래스가 적합해 보여 클래스로 만들었습니다.

jaemuYeo commented on CodeStarterCamp_Week3/Step1/Class/Person.swift:
```diff
@@ -0,0 +1,31 @@
+    }
+    
+    func buyCoffee(coffee: Coffee, from shop: CoffeeShop) {
+        let price = shop.menu[coffee] ?? 0      // 커피숍에 커피 가격
```
상수와 if문 사이에 개행이 있으면 가독성이 좋아질 것 가아요 :)

ollekil replied:
넵 그게 더 가독성이 좋겠네요. 감사합니다.

jaemuYeo commented on CodeStarterCamp_Week3/Step1/Class/Person.swift:
```diff
@@ -0,0 +1,31 @@
+        self.money = money
+    }
+    
+    func buyCoffee(coffee: Coffee, from shop: CoffeeShop) {
```
해당 메서드 네이밍도 파라미터 네임과 겹치는 것 같아요~
buy(coffee:shop)로 지정해도 좋을 것 같습니다!

ollekil replied:
아 이게 더 좋아 보이네요.
파람을 coffe -> coffeeType 으로 수정 했는데 잼킹님 말씀처럼 함수를 buy로 바꾸는게 나아보이네요.

jaemuYeo reviewed (changes requested):
Round3도 금방 수행해주셨군요! 👍
이전에도 전달드렸듯 주석이 많아 가독성을 많이 헤치고있는 것 같아요 ㅠㅠ
코드만으로도 충분히 제 역할들을 나타내고있으니
주석은 제거하시면 좋을 것 같습니다 :)

현재 출력되는 결과가
음료가 나왔으니 픽업해라 -> ㅇㅇ가 음료를 ㅇㅇ원에 샀다. 의 흐름인데
순서가 반대로 되면 흐름이 더 자연스러울 것 같습니다!

ollekil committed:
수정내역
1. 가독성을 해치는 주석 해제
2. 함수명과 매개변수 명이 겹치는 부분 수정
3. 금액을 Double 로 작업한걸 Int 로 변경
4. 커피를 샀다는 출력 함수 위치 수정

ollekil committed:
수정내역
1. 가독성을 해치는 주석 제거

jaemuYeo reviewed (commented):
변경 된 코드 확인하였습니다!!
폴더와 파일 분리를 타입별로 너무 잘해주셨는데
여러 타입과 메서드가 정의되었지만 커밋 이력이 아쉽네요 ㅠㅠ

[카르카 커밋 컨벤션](https://karma-runner.github.io/6.4/dev/git-commit-msg.html)을 참고하여 추후에는 단위별로 커밋해보시면 좋을 것 같아요 :)

5주간 너무너무 고생많으셨습니다. 
저희 교육과정이 많은 도움이 되었으면 좋겠습니다 🙏
이후 남은 챕터와 과제들은 서두르지않고 하나씩 학습해나가시면 될 것 같습니다!
